include/ppx/command_line_parser.h

```
@@ -54,100 +53,114 @@ struct StandardOptions
-            if constexpr (std::is_same_v<T, std::string>) {
-                return value;
-            }
-            else if constexpr (std::is_same_v<T, bool>) {
-                return GetValueAsBool(defaultValue);
-            }
-
-            std::stringstream ss{value};
-            T                 val;
-            ss >> val;
-            if (ss.fail()) {
-                return defaultValue;
-            }
-            return val;
+        return allOptions.find(option) != allOptions.end();
+    }
+
+    size_t GetNumOptions() const { return allOptions.size(); }
+
+    // Try to parse option string into the type of the default value and return it.
+    // If the value fails to be converted, return the specified default value.
+    template <typename T>
+    T GetOptionValueOrDefault(const std::string& optionName, const T& defaultValue) const
+    {
+        if (allOptions.find(optionName) == allOptions.cend()) {
+            return defaultValue;
         }
+        auto valueStr = allOptions.at(optionName).back();
+        return GetParsedOrDefault<T>(valueStr, defaultValue);
+    }
 
-    private:
-        // For boolean options, accept "true" and "false" as well as numeric values.
-        bool GetValueAsBool(bool defaultValue) const
-        {
-            std::stringstream ss{value};
-            bool              val;
-            ss >> val;
-            if (ss.fail()) {
-                ss.clear();
-                ss >> std::boolalpha >> val;
-                if (ss.fail()) {
-                    return defaultValue;
-                }
-            }
-            return val;
+    // Intended for list flags, specified on command line with comma-separated parameters and/or multiple instances of same flag
+    template <typename T1>
+    std::vector<T1> GetOptionValueOrDefault(const std::string& optionName, const std::vector<T1>& defaultValues) const
```

Those 2 functions seems weird:

For non-vector options, you still store all the values, even if only the last will be used. Why store those values since they are never used?

Feels like you should have 2 storages:
 - 1 for options that are specified once (and overridden if specified multiple times), and 1 for options that can be specified multiple times.
 
 For the GetOptionValueOrDefault, since we cannot overload just with the return value, I can imagine this 2 function to still exist. Not sure if having 2 different names is more usable or not. Maybe you and Aliya have another opinion?

---

Isn't that the current aim? I thought this was built to replace the legacy method 😀

---

I'd be in favor to drop the `GetExtraOptions()` and only rely on knobs. Silently fail when there is a typo in a flag is not a great user experience.
But yes, maybe that's too much for 1 PR, so up to you.

Do you plan to keep the unorderd_map in the long run?
Ignoring extra options, with this PR, if I do
`./binary --help=true --help=false --help=true`, you store 3 values for `help`, even if only the last one is required.
Probably not that important if fixing this makes the code less readable, just surprising.